V1.0.0

cat-haines opened the pull request.

cat-haines committed:
Minor refactoring + comments

cat-haines committed:
Added get|setPosition methods

cat-haines committed:
Added README

cat-haines committed:
Updates to improve interface
- Added onComplete callback for readSync
- Added more useful error messages for constructor + change order of params

blindman2k commented:
## The code
- In _disable() you must decrement the _enables counter before not after.
- Consider forcing alignment with a block instead of a sector. It doesn't matter for the logger but it does with the SPIFlashFS. 
- Consider changing all the statics to consts.
- Good use of the new Serialiser.sizeof() but might be worth also checking imp.getfreememory().
- readSync is synchronous so doesn't require an onComplete() callback. If for whatever reason you want to keep it, it should not be called asynchronously.
- To reduce wear on the front of the spi flash we should NOT reset the current position when erasing the storage. Just keep writing forward.

## The readme
- The write(object) documentation refers to a non-existant log(object) function
- You have merged getPosition() with setPosition(sector, offset) into one command getPosition(sector, offset)
- getPosition() and setPosition() could easily work with one number (sector \* SECTOR_SIZE + offset) rather than a table.

cat-haines committed:
Changed _enable and _disable to increment/decrement _enables inline
- _disable() is now properly decrementing _enables before checking

cat-haines committed:
Made statics into constants for performance

cat-haines committed:
Removed onComplete callback from readAsync
- Made onData callback syncronous (could knock device offline for very long reads)

cat-haines committed:
Changed .log to .write in README

cat-haines committed:
Fixed docs for getPosition and setPosition

cat-haines committed:
Changed getPosition and setPosition to use an integer instead of a table
- Also bug fix in _getSectorMetadata

cat-haines commented:
- We aren't going to force alignment with blocks instead of sectors - that decreases the value of this class at the convenience of another, which the developer may or may not be using. If a developer is planning on using both, they simply need to specify appropriate start and end blocks/sectors for the two seperate classes. 
- I don't think we need to check imp.getmemoryfree() - doesn't add much value as we would simply be throwing an error that's no more descriptive than "Out of memory"
- Calling erase() doesn't reset the spiflash pointers..
  - That said, I think there's a bug in the _erase method.. it's setting the internal map variable to `SPIFLASHLOGGER_SECTOR_CLEAN` for the sector we're erasing, but not saving the map into the SPIFlash.. should it be?

cat-haines commented:
Realizing now we don't need to write the map to memory since 0xFFFF is clean, and that's the state of the map after an erase / zeroing!

I think that's all the issues you raised - let me know how the new class looks :)

cat-haines committed:
Removed line that set _next_sec_id in setPosition
- The SPIFlashLogger class knows best.. trying to do this manually causes many very strange errors in readSync (CRC errors, missing data, etc)
